### ISSUE TITLE
Ambiguous types in AddressInfo

### ISSUE THREAD

**Issue by bgamari** [lpeterse/haskell-socket#10]
The `f` type variable in the `GetNameInfo` class definition in `System.Socket.Internal.AddressInfo` appears to be ambiguous,

``` hs
class (Family f) => GetNameInfo f where
  getNameInfo :: SocketAddress f -> NameInfoFlags -> IO (BS.ByteString, BS.ByteString)
```

I'd imagine the intent here was that the first argument would determine `f`, but to do so `SocketAddress` would need to be injective. As far as I know there is no way that GHC would know this. This causes the package to fail to build with GHC 8.0 with,

```
src/System/Socket/Internal/AddressInfo.hsc:303:3: error:
    • Couldn't match type ‘SocketAddress f0’ with ‘SocketAddress f’
      NB: ‘SocketAddress’ is a type function, and may not be injective
      The type variable ‘f0’ is ambiguous
      Expected type: SocketAddress f
                     -> NameInfoFlags -> IO (BS.ByteString, BS.ByteString)
        Actual type: SocketAddress f0
                     -> NameInfoFlags -> IO (BS.ByteString, BS.ByteString)
    • In the ambiguity check for ‘getNameInfo’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      When checking the class method:
        getNameInfo :: forall f.
                       GetNameInfo f =>
                       SocketAddress f
                       -> NameInfoFlags -> IO (BS.ByteString, BS.ByteString)
      In the class declaration for ‘GetNameInfo’
```

Note that GHC 8.0 will include support for [injectivity annotations](http://downloads.haskell.org/~ghc/master/users-guide//glasgow_exts.html?highlight=typefamilydep#ghc-flag--XTypeFamilyDependencies) but unless you want to drop support for earlier compilers you will likely want to either find another way to disambiguate this type or make `SocketAddress` a data family (which are injective).

To see why this code is problematic, consider the following silly example,

``` hs
{-# LANGUAGE TypeFamilies #-}
class Family f where
    type Address f

data Inet6
data Inet7

data AnAddress = AnAddress

instance Family Inet6 where type Address Inet6 = AnAddress
instance Family Inet7 where type Address Inet7 = AnAddress

class (Family f) => GetNameInfo f where
    getNameInfo :: Address f -> String

instance GetNameInfo Inet6 where getNameInfo AnAddress = "Inet6"
instance GetNameInfo Inet7 where getNameInfo AnAddress = "Inet7"

main = putStrLn (getNameInfo AnAddress)
```

What should this print? Neither I nor GHC have any idea since both `Address Inet6` and `Address Inet7` unify with `AnAddress`. The fact that typeclasses are open means that even if there is only one instance in your library, anyone may later come along and add another `Family`, potentially introducing ambiguity.

**Comment by bgamari** [comment]
For the record, the a similar program written with injectivity annotations would like this,

``` hs
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeFamilyDependencies #-}

type family Address f = r | r -> f

data Inet6
data Inet7

data AnAddress = AnAddress

type instance Address Inet6 = AnAddress
--type instance Address Inet7 = AnAddress

class GetNameInfo f where
    getNameInfo :: Address f -> String

instance GetNameInfo Inet6 where getNameInfo AnAddress = "Inet6"
--instance GetNameInfo Inet7 where getNameInfo AnAddress = "Inet7"

main = putStrLn (getNameInfo AnAddress)
```

Moreover, if you uncomment the `Inet7` lines GHC will fail with an error, informing you that you have violated the injectivity requirement of the family.

**Comment by lpeterse** [comment]
Hi Ben,

thanks for noting this and writing such an elaborate example. I'll look into this, maybe at the weekend. I need some time to get into this (again).

I was not aware about upcoming injectivity annotations when I wrote this library :-)

**Comment by lpeterse** [comment]
I followed your proposal and had a look at data families again. Both, injectivity annotation and data families seem to be a viable solution to the issue. At the moment, I tend a bit more towards the data families:

Pro data family:
- Conceptually, the address type wouldn't need to/couldn't exist without the socket family. This would be correctly modeled by using a data family. 
- On the other hand, the socket family can exist and be defined without specifying an address type.
- Supported on older compiler versions.

Pro injectivity annotation:
- Arbitrary (existing) type may be used as address without additional boxing (probably not desired anyway).

``` hs
{-# LANGUAGE TypeFamilies #-}
module Experiment where

class Family f where
data family Address f

class (Family f) => GetNameInfo f where
    getNameInfo :: Address f -> String

data Inet
data instance Address Inet = InetAddress
instance Family Inet where
instance GetNameInfo Inet where
  getNameInfo InetAddress = "INET"

data Inet6
--data instance Address Inet6 = Inet6Address
instance Family Inet6 where
instance GetNameInfo Inet6 where
  getNameInfo _ = "INET6"

main :: IO ()
main = do
  print $ getNameInfo InetAddress
  -- This one surprised me a little... (No data instance!)
  print $ getNameInfo (undefined :: Address Inet6)
```

I'll continue investigating what the implications of switching to data families are.

**Comment by lpeterse** [comment]
Will be fixed in upcoming version `0.6.0.0`.
